### DISCUSSION TITLE
[Proposal] Make event a contextual keyword

### DISCUSSION THREAD

**Discussion by kharai-tor** [dotnet/csharplang#8087] [Language Ideas] [answered]
There are scenarios where "Event" can be a business object, for example an EF entity within a software created to sell tickets or a class in an event logging system.

Here are usages of `@event` in Sentry's .NET SDK for example:
https://github.com/search?q=repo%3Agetsentry%2Fsentry-dotnet%20%40event&type=code

In such scenarios it would be much more pleasant to just be able to name variables and parameters `event` instead of  `@event`.

The `event` keyword appears in such a specific spot in C# grammar that it could easily be a contextual keyword(or so I think) and thus allow the word to be used as an identifier too. In fact, I assume the `event` keyword has been around since before contextual keywords even existed in C# otherwise it might have been one of them right from the start.

**Comment by huoyaoyuan** [comment]
Contextual keywords also existed since the beginning of C#, for the `get`/`set` accessors of properties and `add`/`remove` properties of events.

The position of `event` keyword is not an easily distinguishable context. It appears like a modifier of a field. All modifiers are not contextual so that they are clearly distinguishable with type name.

**Reply by kharai-tor** [reply]
So with these two examples:

``` C#
event e;

event Action e;
```

The former could be interpreted as a type name and the latter could be interpreted as a keyword. So they could be distinguished by the number of words basically. And it seems other keywords can't appear after the `event` keyword so if that's true the ordering of keywords shouldn't mess with that.

Am I missing something?

**Reply by huoyaoyuan** [reply]
> So they could be distinguished by the number of words basically.

It may be in theory distinguishable, but troublesome for compile and IDE to support, especially when you are typing partial code.

**Reply by HaloFour** [reply]
From a parsing perspective I imagine it wouldn't be at all difficult to make the `event` keyword contextual.  Question is whether there's a lot of value in doing so, especially since it's easy to quote it with `@`.  There seems to be more interest in having the compiler be a bit more opinionated about what is a keyword vs. what should be a type name, so this proposal feels like a step in the opposite direction.

**Reply by kharai-tor** [reply]
The whole thing with contextual keywords, if I understand correctly, is to allow identifiers as much as possible at the cost of making the implementation more difficult. You could have all contextual keywords as regular keywords, make the implementation easier and be able to escape them all with `@`. But since that's not the case and the notion of contextual keywords exists I felt that `event` fit in that bucket. But again, that might just be me. If no one agrees I'll close the discussion.

**Reply by davidroth** [reply]
You are not alone. Been in this situation many times and als find it ugly to use `@event`.
I remember that I sometimes used `e` instead.

**Comment by CyrusNajmabadi** [comment]
Why is @event unpleasant?

**Reply by kharai-tor** [reply]
Well that could just be me. I'll leave this proposal open for a bit. If no one agrees I'm happy to close it. Definitely don't mean to annoy anybody.

**Reply by CyrusNajmabadi** [reply]
>  Definitely don't mean to annoy anybody.

You haven't annoyed anyone :)  

These are just questions meant to understand your position better :)

**Reply by HaloFour** [reply]
Never hurts to have the conversation.  IMO, of the keywords `event` is probably of the easiest to make contextual given that it can only be used as a keyword in fairly specific circumstances.  I don't think it would create any ambiguity to make it contextual with existing syntax.  There might be a question if it would limit future evolution to do so.

The contextual keyword conversation I mentioned above I think more applies to the cases where a feature was added that could be ambiguous with existing legal syntax, and the language went out of its way to avoid breaking changes.  For instance, `var` is only a keyword if a type named `var` is not in scope.  This creates warts and friction in the language, and at this point there the thought that it's better to accept the break to simplify the language going forward.  I don't think `event` would fit into that category.  At that point the question is whether it's worth the effort to do so.

**Reply by kharai-tor** [reply]
@CyrusNajmabadi 
> These are just questions meant to understand your position better :)

I've had to type it enough to have wished for it on various separate occasions. I even tried to pull Roslyn and implement it myself a couple years back in my free time just to see if I could get it done. I couldn't. So I understand the difficulty and reluctance to implement it at least to a degree. In the end I thought I'd try my luck and mention it anyway.

@HaloFour
> There might be a question if it would limit future evolution to do so.

I see your point.

**Reply by The-Futurist** [reply]
@TKharaishvili 
The `event` keyword appears in the C# grammar [here](https://github.com/antlr/grammars-v4/blob/master/csharp/CSharpParser.g4), but this is old, only supports C# 6. You could experiment by adjusting the grammar to allow an identifier to be named "event" but still correctly parse the other existing uses of the term. If you can do that then that shows that in principle this could be done.

**Reply by CyrusNajmabadi** [reply]
@Korporal  our up to date grammar is in this repo here: https://github.com/dotnet/roslyn/blob/main/src/Compilers/CSharp/Portable/Generated/CSharp.Generated.g4.  It is always kept in sync with whatever version of the language the compiler supports in whatever branch you are looking in.

**Reply by The-Futurist** [reply]
Oh, that's interesting I somehow had the impression you don't use Antlr. Or do you have your own tools and _generate_ that .g4 file?

**Reply by The-Futurist** [reply]
Hmm, running that through antlr4 gens a lot of errors...

![image](https://github.com/dotnet/csharplang/assets/12262952/81cfcd61-96c2-4dab-b9eb-0cb84d8fd5fc)

**Reply by CyrusNajmabadi** [reply]
That is all accurate. There are sections of the grammar that would need to be filled in. (The lexical grammar portions).  The syntactic grammar is complete.

**Reply by iam3yal** [reply]
@CyrusNajmabadi 

> Why is @event unpleasant?

I have a tiny, tiny issue with it when selecting it, VS and VSCode seems to select `event` as opposed to `@event` but then I guess that this is a tooling issue that should be addressed as opposed to a language issue.

**Reply by CyrusNajmabadi** [reply]
Yes.  please file issue if that's happening.

**Reply by The-Futurist** [reply]
> That is all accurate. There are sections of the grammar that would need to be filled in. (The lexical grammar portions). The syntactic grammar is complete.

So what is the file for? why do they go to the trouble of generating `CSharp.Generated.g4`? what's it used for?

**Reply by CyrusNajmabadi** [reply]
To provide a grammar for the language. For whatever purpose someone might see fit to use it. 

Also, as I mentioned above, in a very real way, it clearly documents the grammar that each branch/release of ours supports. For example, if you go into the 'features/roles' branch, you'll see the updated grammar for 'extension types'.  

This also helps ensure that as we merge prs, or features back into 'main' that the language changes are easy to review and in line with expectations.

**Reply by The-Futurist** [reply]
OK, I'm still not seeing why the file is created though, seems it can only be consumed by a human reader. Why exclude the lexical definitions anyway? you might make a change at the lexical level and surely seeing that in the file would be helpful. Is this all a Roslyn question, if so I might raise it there.

**Reply by HaloFour** [reply]
I don't know if there are newer conversations.  IIRC the antlr file is only there for reference, and the team has no interest in going through the motions to create and maintain a second reference parser.

https://github.com/dotnet/roslyn/issues/8379
https://github.com/dotnet/csharplang/issues/1902

**Reply by CyrusNajmabadi** [reply]
> OK, I'm still not seeing why the file is created though, 

I answered this already, with several reasons given. 

> Why exclude the lexical definitions anyway?

I opened a pr to add them. 

> you might make a change at the lexical level and surely seeing that in the file would be helpful.

Agreed. 

> Is this all a Roslyn question, if so I might raise it there.

Feel free.

**Comment by erik-kallen** [comment]
YES, Please.

I use events (as in the theoretical concept) all the time. I use events (as in the C# keyword) pretty much never. Being able to name my "events" `event` instead of `evt` would be nice.

**Reply by CyrusNajmabadi** [reply]
You can use `event` as a name today.  You just need to prefix with `@`.  C# has always supported using any keyword in the language in this fashion precisely so you don't have to make odd abbrs for those words.

**Reply by The-Futurist** [reply]
I wonder how many C# developer though would _prefer_ to use the name `@event` rather than `event`? I'd venture to say none.

**Reply by CyrusNajmabadi** [reply]
That doesn't really change anything.  We have a solution for this issue.  It's been in the lang since 1.0.  In 25 years, we haven't relaxed a single keyword.   I don't see that happening now for a single keyword just because some people would prefer to leave off an `@`.  Fundamentally, this saves a character, at the cost of more complexity and time spent away from high value feature work.  There's no realistic chance of this happening.

**Reply by huoyaoyuan** [reply]
I've never had a requirement to use `@event`. My events are always named like with their meaning like properties.

**Reply by erik-kallen** [reply]
Sure, I can use `@event`. But it's even more ugly than `evt` so it doesn't really improve anything (IMO).

**Reply by The-Futurist** [reply]
@erik-kallen 
> Sure, I can use `@event`. But it's even more ugly than `evt` so it doesn't really improve anything (IMO).

I'm afraid you're stuck, as you can read the cost of changing the system to support this is deemed to far outweigh the benefits. Even though in principle this could be changed it's not so easy that the cost is trivial. Some grammars are designed specifically to not restrict identifiers like this but not C, C++, Java, C# and so on, restricting identifiers this way is somewhat old fashioned and [causes problems](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/warning-waves#cs8981---the-type-name-only-contains-lower-cased-ascii-characters) when new keywords are added in case that new keyword clashes with already used identifiers.

**Reply by CyrusNajmabadi** [reply]
The cost is marginal here.  But i don't see the benefit.  We have an existing solution to this.  I don't see the need for two solutions.

**Reply by CyrusNajmabadi** [reply]
Remember that every feature starts at -100.  A feature has to actually justify itself.  So far, that's not bee present here.

**Reply by HaloFour** [reply]
> as you can read the cost of changing the system to support this is deemed to far outweigh the benefits.

In this case the cost is relatively trivial, but the benefits are even more trivial, thus it isn't worth the effort.

> Some grammars are designed specifically to not restrict identifiers like this

They trade one restriction for another.  But if you find languages like that so much better to use, feel free to go use them.  You know that C# will never be such a language, so why keep harping on it ad nauseam?

> when new keywords are added in case that new keyword clashes with already used identifiers.

Which isn't relevant to this conversation as `event` has been a keyword since C# 1.0.

**Reply by The-Futurist** [reply]
> You know that C# will never be such a language, so why keep harping on it ad nauseam?

Why the rudeness? my remarks were addressed to @erik-kallen, why do you have to complain all the time?

**Reply by CyrusNajmabadi** [reply]
@Korporal please keep things on topic.  Halo's responses were about the costs and the continued digressions into non-c# grammars that seem to commonly come with your posts.  This is the repo for C# discussion.  The aspects of other languages isn't relevant here as we're not trying to follow what others do, but set our own rules and designs based on what we think is best for the ecosystem, including revisiting and relitigating decisions from 25 years ago that we haven't seen any need to change.

**Comment by The-Futurist** [comment]
> Remember that every feature starts at -100. A feature has to actually justify itself. So far, that's not bee present here.

If the cost of some feature is _known_ to be "marginal" then shouldn't that feature start at -1 or something?

**Reply by 333fred** [reply]
No. The smallest "marginal" feature will still cost dozens of hours of engineering time, just in approving it in LDM, implementing, testing, adding new documentation, etc.

**Reply by CyrusNajmabadi** [reply]
No.  That's the entire point.  Features have to deliver substantive value.  That value has to overcome the fact that any feature comes, by default, with *detrimental* impact to the ecosystem and community, regardless of the cost to implement it.

So the *feature* itself must outweigh those detriments.  Often *substantively* for us to take it.

**Comment by The-Futurist** [comment]
> So the _feature_ itself must outweigh those detriments. Often _substantively_ for us to take it.

I was agreeing with you, I wrote "the cost of changing the system to support this is deemed to far outweigh the benefits",

**Reply by CyrusNajmabadi** [reply]
the cost to change the system here is marginal.  The gain though is even less.    The feature *starts* at -100.  We want to see real widespread benefit to warrant even considering it (even if the costs are tiny).  We reject the majority of proposals here simply because the feature doesn't meet that hurdle and doesn't get championed thusly.

Also, can you please actually reply to the threads, and not start new threads :)

**Reply by The-Futurist** [reply]
> the cost to change the system here is marginal. The gain though is even **less**. 
 
 Indeed so the cost is greater than the benefit.

**Reply by CyrusNajmabadi** [reply]
@Korporal I'm not saying it's cost vs benefit here.  I'm saying, it's "benefit" *period*.   The benefit isn't here *at all* regardless of cost.  This could cost *zero* and i wouldn't take it because of detriment to the ecosystem.   :)

> as you can read the cost of changing the system

**Comment by CyrusNajmabadi** [comment, marked as answer]
We are not going to get into digressions about other languages, or minutia of cost analyses.  The position for us on keywords so far is that we think they are valuable for the language and ecosystem, and that changing existing keywords to be contextual is a non-goal.  If that ever changes, we'd revisit this.  But for now, this has effectively been subsumed into the decisions in this space we've had several times over the lifetime of C#.
